core/trino-main/src/main/java/io/trino/sql/planner/optimizations/DeterminePartitionCount.java

```diff
@@ -0,0 +1,247 @@
+        }
+
+        return plan;
+    }
+
+    private Optional<Integer> determinePartitionCount(PlanNode plan, Session session, TypeProvider types, TableStatsProvider tableStatsProvider)
+    {
+        int maxHashPartitionCount = getMaxHashPartitionCount(session);
+        long minInputRowsPerTask = getMinInputRowsPerTask(session);
+
+        if (minInputRowsPerTask == 0) {
+            return Optional.empty();
+        }
+
+        // Skip for expanding plan nodes like CROSS JOIN or UNNEST which can substantially increase the amount of data.
+        if (isInputMultiplyingPlanNodesPresent(plan)) {
+            return Optional.empty();
+        }
+
+        CachingStatsProvider statsProvider = new CachingStatsProvider(statsCalculator, session, types, tableStatsProvider);
+
+        double sourceTablesRowCount = getSourcePlanNodesRowCount(plan, statsProvider);
+        double expandingNodesMaxRowCount = getExpandingNodesMaxRowCount(plan, statsProvider);
+        if (isNaN(sourceTablesRowCount) || isNaN(expandingNodesMaxRowCount)) {
+            return Optional.empty();
+        }
+
+        int partitionCount = max(
+                getPartitionsFromRowCount(sourceTablesRowCount, minInputRowsPerTask),
+                getPartitionsFromRowCount(expandingNodesMaxRowCount, minInputRowsPerTask));
+
+        if (partitionCount >= maxHashPartitionCount) {
+            return Optional.empty();
+        }
+
+        log.debug("Estimated remote exchange partition count for query %s is %s", session.getQueryId(), partitionCount);
+        return Optional.of(partitionCount);
+    }
+
+    private static int getPartitionsFromRowCount(double rowCount, long minInputRowsPerTask)
+    {
+        return max((int) (rowCount / minInputRowsPerTask), 1);
+    }
+
+    private static boolean isInputMultiplyingPlanNodesPresent(PlanNode root)
+    {
+        return PlanNodeSearcher.searchFrom(root)
+                .where(DeterminePartitionCount::isInputMultiplyingPlanNode)
+                .matches();
+    }
+
+    private static boolean isInputMultiplyingPlanNode(PlanNode node)
+    {
+        if (node instanceof UnnestNode) {
+            return true;
+        }
+
+        if (node instanceof JoinNode crossJoinNode && ((JoinNode) node).isCrossJoin()) {
+            // If any of the input node is scalar then there's no need to skip cross join
+            return !isAtMostScalar(crossJoinNode.getRight()) && !isAtMostScalar(crossJoinNode.getLeft());
+        }
+
+        return false;
+    }
+
+    private static double getExpandingNodesMaxRowCount(PlanNode root, StatsProvider statsProvider)
+    {
+        List<PlanNode> expandingNodes = PlanNodeSearcher.searchFrom(root)
+                .where(DeterminePartitionCount::isExpandingPlanNode)
+                .findAll();
+
+        return expandingNodes.stream()
+                .mapToDouble(node -> statsProvider.getStats(node).getOutputRowCount())
+                .max()
+                .orElse(0);
```

Comment:
`Integer.MAX_VALUE`? It should be safer to overestimate.

Reply:
This is not required since `orElse` will only called when `expandingNodes` is empty that basically means there are not expanding nodes.